File: backend/app/schemas/activity.py
from __future__ import annotations
import uuid
from datetime import datetime, date
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.db.models import ActivityStatus
from app.schemas.user import UserResponse


class ActivityCreate(BaseModel):
    client_name: str
    entry_date: date
    action_item: str
    assigned_to_id: Optional[uuid.UUID] = None
    product_domain: Optional[str] = None
    activity_type: Optional[str] = None
    target_closure_date: Optional[date] = None
    actual_closure_date: Optional[date] = None
    status: ActivityStatus = ActivityStatus.open
    remarks: Optional[str] = None


class ActivityUpdate(BaseModel):
    client_name: Optional[str] = None
    entry_date: Optional[date] = None
    action_item: Optional[str] = None
    assigned_to_id: Optional[uuid.UUID] = None
    product_domain: Optional[str] = None
    activity_type: Optional[str] = None
    target_closure_date: Optional[date] = None
    actual_closure_date: Optional[date] = None
    status: Optional[ActivityStatus] = None
    remarks: Optional[str] = None
    target_date_change_remarks: Optional[str] = None


class ActivityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    client_name: str
    entry_date: date
    action_item: str
    assigned_to_id: Optional[uuid.UUID]
    assigned_to: Optional[UserResponse] = None
    product_domain: Optional[str]
    activity_type: Optional[str]
    target_closure_date: Optional[date]
    initial_target_closure_date: Optional[date] = None
    actual_closure_date: Optional[date]
    status: ActivityStatus
    remarks: Optional[str]
    created_by_id: Optional[uuid.UUID]
    created_by: Optional[UserResponse] = None
    created_at: datetime
    updated_at: datetime


class ActivityListResponse(BaseModel):
    items: list[ActivityResponse]
    total: int
    page: int
    page_size: int


class RejectRequest(BaseModel):
    reason: Optional[str] = None


class TargetDateHistoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    activity_id: uuid.UUID
    old_date: Optional[date]
    new_date: Optional[date]
    remarks: Optional[str]
    changed_by: Optional[UserResponse] = None
    changed_at: datetime

File: frontend/src/@types/activity.ts
import { User } from "./user";

export type ActivityStatus = "Open" | "In Progress" | "Closed" | "On-Hold" | "Completed";

export interface TargetDateHistory {
  id: string;
  activity_id: string;
  old_date: string | null;
  new_date: string | null;
  remarks: string | null;
  changed_by: User | null;
  changed_at: string;
}

export interface Activity {
  id: string;
  client_name: string;
  entry_date: string;
  action_item: string;
  assigned_to_id: string | null;
  assigned_to?: User | null;
  product_domain: string | null;
  activity_type: string | null;
  target_closure_date: string | null;
  initial_target_closure_date: string | null;
  actual_closure_date: string | null;
  status: ActivityStatus;
  remarks: string | null;
  created_by_id: string | null;
  created_by?: User | null;
  created_at: string;
  updated_at: string;
}

export interface ActivityCreate {
  client_name: string;
  entry_date: string;
  action_item: string;
  assigned_to_id?: string | null;
  product_domain?: string | null;
  activity_type?: string | null;
  target_closure_date?: string | null;
  actual_closure_date?: string | null;
  status: ActivityStatus;
  remarks?: string | null;
}

export interface ActivityUpdate extends Partial<ActivityCreate> {
  target_date_change_remarks?: string | null;
}

export interface ActivityListResponse {
  items: Activity[];
  total: number;
  page: number;
  page_size: number;
}

export const ACTIVITY_STATUSES: ActivityStatus[] = [
  "Open", "In Progress", "Closed", "On-Hold", "Completed"
];

export const ACTIVITY_TYPES: string[] = [
  "Offshore", "Documentation", "Project", "Resource", "Follow Up",
  "Proposal", "R&D", "POC", "Analysis", "Support", "Internal",
  "Effort Estimation", "Knowledge Transfer", "Training", "Development",
  "Testing", "Webinar", "CR", "Audit", "Other",
];

export const PRODUCT_DOMAINS: string[] = [
  "RPA", "Full Stack", "DevOps", "Mendix", "AI", "Mobile", ".Net",
  "Resource Augmentation", "Data Engineering", "Power Platform",
  "Internal", "PMO", "After Market", "Teamcenter", "AI & ML",
  "Python Full Stack", "Data Science", "Automation", "DBA",
  "Performance Testing", "Polarion", "CMS", "Lotus Notes", "Other",
];
